src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +238,813 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+            # integer values for these Literals
+            sympy_ref1s = \
+                SymPyWriter.convert_to_sympy_expressions(ref1_accesses)
+
+            # We have a set of Literal values, we use the SymPyWriter to
+            # convert these objects to expressions we can use to obtain
+            # integer values for these Literals
+            sympy_ref2s = \
+                SymPyWriter.convert_to_sympy_expressions(ref2_accesses)
+            return self._check_valid_overlap(sympy_ref1s, sympy_ref2s)
+
+        return True
+
+    def _check_dependency_pairing_valid(self, node1, node2, task1, task2):
+        '''
+        Given a pair of nodes which are children of a OMPDependClause, this
+        function checks whether the described dependence is correctly
+        described by the OpenMP standard.
+        If the dependence is not going to be handled safely, this function
+        returns False, else it returns true.
+
+        :param node1: the first input node to check.
+        :type node1: :py:class:`psyclone.psyir.nodes.Node`
+        :param node2: the second input node to check.
+        :type node2: :py:class:`psyclone.psyir.nodes.Node`
+        :param task1: the OMPTaskDirective node containing node1 as a \
+                      dependency
+        :type task1: :py:class:`psyclone.psyir.nodes.OMPTaskDirective`
+        :param task2: the OMPTaskDirective node containing node2 as a \
+                      dependency
+        :type task2: :py:class:`psyclone.psyir.nodes.OMPTaskDirective`
+
+        :returns: whether the dependence is going to be handled safely \
+                  according to the OpenMP standard.
+        :rtype: bool
+        '''
+        # Checking the symbol is the same works. If the symbol is not the same
+        # then there's no dependence, so its valid.
+        if node1.symbol != node2.symbol:
+            return True
+        # For structure reference we need to check they access
+        # the same member. If they don't, no dependence so valid.
+        if isinstance(node1, StructureReference):
+            # If either is a StructureReference here they must both be,
+            # as they access the same symbol.
+            member = node1
+            member1 = node2
+
+            # We can't just do == on the Member child, as that
+            # will recurse and check the array indices for any
+            # ArrayMixin children
+
+            # Check the signature of both StructureReference
+            # to see if they are accessing the same data
+            ref0_sig = node1.get_signature_and_indices()[0]
+            ref1_sig = node2.get_signature_and_indices()[0]
+            if ref0_sig != ref1_sig:
+                return True
+
+        # If we have (exactly) Reference objects we filter out
+        # non-matching ones with the symbol check, and matching ones
+        # are always valid since they are simple accesses.
+        if type(node1) is Reference:
```

Comment:
I think no.

I think the check of `if node1.symbol != node2.symbol` means that `type(node1) == type(node2)` by construction, so checking both is not needed - @arporter @sergisiso is my assumption here correct?

Reply:
I don't have the context but if they both have a `symbol` property then they must both be References of some sort. However, they *could* be different types of Reference (e.g `ArrayReference` and `Reference`).

Reply:
Is this down to the `Reference` class being generic again? I think all the members of a Depend clause in a normal usecase will be safe for my assumption - but I could change the original test to
```
if node1.symbol != node2.symbol or type(node1) != type(node2):
```
which would handle that edge case - I think this again hits the generic `Reference` class not being abstract, but its easy enough to work with at the moment. I'll make that change as I expect it is relatively low cost to avoid any weirdness.

Reply:
Done